Common/CodeBlock.h

```diff
@@ -38,36 +38,65 @@ class CodeBlockCommon {
 	size_t region_size;
 };
 
-template<class T> class CodeBlock : public CodeBlockCommon, public T, NonCopyable
-{
+template<class T> class CodeBlock : public CodeBlockCommon, public T, NonCopyable {
 private:
 	// A privately used function to set the executable RAM space to something invalid.
 	// For debugging usefulness it should be used to set the RAM to a host specific breakpoint instruction
 	virtual void PoisonMemory() = 0;
 
 public:
-	CodeBlock() {}
+	CodeBlock() : writeStart_(nullptr) {}
 	virtual ~CodeBlock() { if (region) FreeCodeSpace(); }
 
 	// Call this before you generate any code.
-	void AllocCodeSpace(int size)
-	{
+	void AllocCodeSpace(int size) {
 		region_size = size;
 		region = (u8*)AllocateExecutableMemory(region_size);
 		T::SetCodePointer(region);
+		// On W^X platforms, we start with writable but not executable pages.
+		if (PlatformIsWXExclusive()) {
+			ProtectMemoryPages(region, region_size, MEM_PROT_READ | MEM_PROT_WRITE);
+		}
 	}
 
 	// Always clear code space with breakpoints, so that if someone accidentally executes
 	// uninitialized, it just breaks into the debugger.
-	void ClearCodeSpace()
-	{
+	void ClearCodeSpace() {
+		if (PlatformIsWXExclusive()) {
+			ProtectMemoryPages(region, region_size, MEM_PROT_READ | MEM_PROT_WRITE);
+		} else {
+			ProtectMemoryPages(region, region_size, MEM_PROT_READ | MEM_PROT_WRITE | MEM_PROT_EXEC);
+		}
 		PoisonMemory();
 		ResetCodePtr();
 	}
 
+	// BeginWrite/EndWrite assumes that we keep appending. If you don't specify a size and we encounter later executable block, we're screwed.
+	// These CANNOT be nested.
+	void BeginWrite(size_t sizeEstimate = 1) {
```

Comment:
Hmm... won't this just give us a page (4096 bytes) by default?  I'm not sure anything limits DoJit to 4096 bytes of code specifically per block - although I think we do have a limit, it may be higher.  I believe I hit it once in Gods Eater Burst.

Maybe it would be better to default to -1, and then consider -1 to mean start to end?  Or a higher amount than 4k?

Or are we assuming that while DoJit'ing, the memory ahead will be writable until we need to mark it executable, and then when we clear, it'll be reset to writable again?  If so, probably deserves a comment that we're relying on that behavior...

-[Unknown]